library/widget/tactic_component.lean

```diff
@@ -0,0 +1,92 @@
+/-
+Copyright (c) E.W.Ayers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: E.W.Ayers
+-/
+import .basic
+
+meta def tactic.run_simple {α} : tactic_state → tactic α → option α
+| ts t := match t ts with
+          | (interaction_monad.result.success a ts') := some a
+          | (interaction_monad.result.exception _ _ _) := none
+          end
+
+namespace widget
+
+/-- A component that implicitly depends on tactic_state. For efficiency we always assume that the tactic_state is unchanged between component renderings. -/
```

Comment:
> efficiency

Can you elaborate?

Reply:
otherwise you have to check whether the entire tactic state has changed which you can't do without implementing `hash` for tactic_state objects.

Reply:
Oh, that's a bummer.  But also a good reason.